File: db4on/tutorial/db4o-tutorial-chapters/src/com/db4o/f1/chapter1/NQExample.cs
using com.db4o;
using com.db4o.query;
using com.db4o.f1;

namespace com.db4o.f1.chapter1
{
	public class NQExample : Util
	{
		public static void Main(string[] args)
		{
			ObjectContainer db = Db4o.OpenFile(Util.YapFileName);
			try
			{
				StorePilots(db);
				RetrieveComplexSODA(db);
				RetrieveComplexNQ(db);
				RetrieveArbitraryCodeNQ(db);
				ClearDatabase(db);
			}
			finally
			{
				db.Close();
			}
		}
    
		public static void StorePilots(ObjectContainer db)
		{
			db.Set(new Pilot("Michael Schumacher", 100));
			db.Set(new Pilot("Rubens Barrichello", 99));
		}
    
		public static void RetrieveComplexSODA(ObjectContainer db)
		{
			Query query=db.Query();
			query.Constrain(typeof(Pilot));
			Query pointQuery=query.Descend("_points");
			query.Descend("_name").Constrain("Rubens Barrichello")
				.Or(pointQuery.Constrain(99).Greater()
				.And(pointQuery.Constrain(199).Smaller()));
			ObjectSet result=query.Execute();
			ListResult(result);
		}

		public static void RetrieveComplexNQ(ObjectContainer db)
		{
			ObjectSet result = db.Query(new ComplexQuery());
			ListResult(result);
		}

		public static void RetrieveArbitraryCodeNQ(ObjectContainer db)
		{
			ObjectSet result = db.Query(new ArbitraryQuery(new int[]{1,100}));
			ListResult(result);
		}
    
		public static void ClearDatabase(ObjectContainer db)
		{
			ObjectSet result = db.Get(typeof(Pilot));
			while (result.HasNext())
			{
				db.Delete(result.Next());
			}
		}
	}
}

File: db4on/tutorial/vb/db4o-tutorial-chapters/src/com/db4o/f1/chapter1/NQExample.vb
Imports com.db4o
Imports com.db4o.query

Namespace com.db4o.f1.chapter1
	Public Class NQExample
	Inherits Util
		Public Shared Sub Main(ByVal args As String())
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
			Try
				StorePilots(db)
				RetrieveComplexSODA(db)
				RetrieveComplexNQ(db)
				RetrieveArbitraryCodeNQ(db)
				ClearDatabase(db)
			Finally
				db.Close()
			End Try
		End Sub

		Public Shared Sub StorePilots(ByVal db As ObjectContainer)
			db.[Set](New Pilot("Michael Schumacher", 100))
			db.[Set](New Pilot("Rubens Barrichello", 99))
		End Sub

		Public Shared Sub RetrieveComplexSODA(ByVal db As ObjectContainer)
			Dim query As Query = db.Query()
			query.Constrain(GetType(Pilot))
			Dim pointQuery As Query = query.Descend("_points")
			query.Descend("_name").Constrain("Rubens Barrichello").[Or](pointQuery.Constrain(99).Greater().[And](pointQuery.Constrain(199).Smaller()))
			Dim result As ObjectSet = query.Execute()
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveComplexNQ(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.Query(New ComplexQuery())
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveArbitraryCodeNQ(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.Query(New ArbitraryQuery(New Integer() {1, 100}))
			ListResult(result)
		End Sub

		Public Shared Sub ClearDatabase(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](GetType(Pilot))
			While result.HasNext()
				db.Delete(result.[Next]())
			End While
		End Sub

	End Class
End Namespace
